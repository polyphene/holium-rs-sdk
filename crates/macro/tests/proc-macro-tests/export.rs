#![allow(unreachable_code)]
use holium_rs_sdk::*;
use serde::{Deserialize, Serialize};

#[holium_bindgen]
pub struct GoodStruct {
    number: u32,
}

#[holium_bindgen]
pub fn pass1() {}

#[holium_bindgen]
pub fn pass2() -> u32 {
    0
}

#[holium_bindgen]
pub fn pass3(a: u32) -> u32 {
    a
}

#[holium_bindgen]
pub fn pass4(a: u32) -> GoodStruct {
    GoodStruct { number: a }
}

#[holium_bindgen]
pub fn pass5(a: GoodStruct) -> GoodStruct {
    a
}

#[holium_bindgen]
pub fn pass6(a: &GoodStruct) -> GoodStruct {
    GoodStruct { number: a.number }
}

#[holium_bindgen]
pub fn pass7(a: &mut GoodStruct) -> GoodStruct {
    a.number += 10;
    let new_struct = GoodStruct { number: a.number };
    return new_struct;
}

#[holium_bindgen]
pub fn pass8(a: &mut GoodStruct) -> (GoodStruct, u32) {
    let old_number = a.number;
    a.number += 10;
    let new_struct = GoodStruct { number: a.number };
    return (new_struct, old_number);
}

#[holium_bindgen]
pub fn pass9(a: &mut GoodStruct, b: String) -> (GoodStruct, String) {
    a.number += 10;
    let new_struct = GoodStruct { number: a.number };
    return (new_struct, b);
}

#[holium_bindgen]
pub fn pass10(a: Option<u32>) -> Option<u32> {
    a
}

#[holium_bindgen]
pub fn pass11(a: Vec<u32>) -> Vec<u32> {
    a
}

struct BadStructNoMacro {
    number: u32,
}

#[holium_bindgen]
pub fn fail1(a: BadStructNoMacro) -> BadStructNoMacro {
    a
}

#[derive(Serialize, Deserialize)]
struct BadStructOnlySerde {
    number: u32,
}

#[holium_bindgen]
pub fn fail2(a: BadStructOnlySerde) -> BadStructOnlySerde {
    a
}

#[holium_bindgen]
pub fn fail3<'a>(x: &'a GoodStruct, y: &'a GoodStruct) -> &'a GoodStruct {
    GoodStruct {
        number: x.number + y.number,
    }
}

#[holium_bindgen]
pub fn fail4<T>(x: T) -> T {
    x
}

fn main() {}
