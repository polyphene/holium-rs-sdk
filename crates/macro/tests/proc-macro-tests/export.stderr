error: can't #[holium_bindgen] functions with lifetime or type parameters
  --> $DIR/export.rs:90:13
   |
90 | pub fn fail3<'a>(x: &'a GoodStruct, y: &'a GoodStruct) -> &'a GoodStruct {
   |             ^^^^

error: can't #[holium_bindgen] functions with lifetime or type parameters
  --> $DIR/export.rs:97:13
   |
97 | pub fn fail4<T>(x: T) -> T {
   |             ^^^

error[E0277]: the trait bound `BadStructNoMacro: Serialize` is not satisfied
   --> $DIR/export.rs:74:1
    |
74  | #[holium_bindgen]
    | ^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `BadStructNoMacro`
    |
   ::: C:/Users/thoma/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_cbor-0.11.2/src/value/ser.rs:440:8
    |
440 |     T: Serialize,
    |        --------- required by this bound in `to_value`
    |
    = note: required because of the requirements on the impl of `Serialize` for `Vec<BadStructNoMacro>`
    = note: this error originates in the macro `vec` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `generate_node` found for struct `BadStructNoMacro` in the current scope
  --> $DIR/export.rs:74:1
   |
70 | struct BadStructNoMacro {
   | ----------------------- function or associated item `generate_node` not found for this
...
74 | #[holium_bindgen]
   | ^^^^^^^^^^^^^^^^^ function or associated item not found in `BadStructNoMacro`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `generate_node`, perhaps you need to implement it:
           candidate #1: `holium_rust_sdk::GenerateNode`
   = note: this error originates in the attribute macro `holium_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BadStructNoMacro: Serialize` is not satisfied
  --> $DIR/export.rs:74:1
   |
74 | #[holium_bindgen]
   | ^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `BadStructNoMacro`
   |
   = note: required by `_::_serde::ser::SerializeStruct::serialize_field`
   = note: this error originates in the derive macro `holium_rust_sdk::internal::serde::Serialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BadStructNoMacro: Deserialize<'_>` is not satisfied
  --> $DIR/export.rs:74:1
   |
74 | #[holium_bindgen]
   | ^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `BadStructNoMacro`
   |
   = note: required by `next_element`
   = note: this error originates in the attribute macro `holium_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `BadStructNoMacro: Deserialize<'_>` is not satisfied
  --> $DIR/export.rs:74:1
   |
74 | #[holium_bindgen]
   | ^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `BadStructNoMacro`
   |
   = note: required by `next_value`
   = note: this error originates in the attribute macro `holium_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `generate_node` found for struct `BadStructOnlySerde` in the current scope
  --> $DIR/export.rs:84:1
   |
80 | struct BadStructOnlySerde {
   | ------------------------- function or associated item `generate_node` not found for this
...
84 | #[holium_bindgen]
   | ^^^^^^^^^^^^^^^^^ function or associated item not found in `BadStructOnlySerde`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `generate_node`, perhaps you need to implement it:
           candidate #1: `holium_rust_sdk::GenerateNode`
   = note: this error originates in the attribute macro `holium_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)
